File: java/src/main/java/com/hydrapayments/sdk/util/SigningUtil.java
package com.hydrapayments.sdk.util;

import javax.crypto.Mac;
import javax.crypto.spec.SecretKeySpec;
import java.security.MessageDigest;
import java.util.Base64;

public final class SigningUtil {

    private static final String HMAC_ALGORITHM = "HmacSHA256";

    private SigningUtil() {}

    public static String signMessage(String secret, String message) {
        if (secret == null || secret.isEmpty()) {
            return "";
        }
        try {
            Mac mac = Mac.getInstance(HMAC_ALGORITHM);
            SecretKeySpec keySpec = new SecretKeySpec(secret.getBytes("UTF-8"), HMAC_ALGORITHM);
            mac.init(keySpec);
            byte[] rawHmac = mac.doFinal(message.getBytes("UTF-8"));
            return Base64.getEncoder().encodeToString(rawHmac);
        } catch (Exception e) {
            throw new RuntimeException("Failed to sign message", e);
        }
    }

    public static boolean verifySignature(String secret, String payload, String signature) {
        if (secret == null || secret.isEmpty()) {
            return false;
        }
        String expected = signMessage(secret, payload);
        try {
            byte[] expectedBytes = Base64.getDecoder().decode(expected);
            byte[] givenBytes = Base64.getDecoder().decode(signature);
            if (expectedBytes.length != givenBytes.length) {
                return false;
            }
            return MessageDigest.isEqual(expectedBytes, givenBytes);
        } catch (IllegalArgumentException e) {
            return false;
        }
    }

    public static String buildSigningMessage(String method, String path, String timestamp, String body) {
        return method + ":" + path + ":" + timestamp + ":" + body;
    }
}

File: android/src/main/kotlin/com/hydrapayments/sdk/SigningUtil.kt
package com.hydrapayments.sdk

import java.util.Base64
import javax.crypto.Mac
import javax.crypto.spec.SecretKeySpec

object SigningUtil {
    fun sign(secret: String, message: String): String {
        if (secret.isEmpty()) return ""
        val mac = Mac.getInstance("HmacSHA256")
        mac.init(SecretKeySpec(secret.toByteArray(Charsets.UTF_8), "HmacSHA256"))
        return Base64.getEncoder().encodeToString(mac.doFinal(message.toByteArray(Charsets.UTF_8)))
    }

    fun verify(secret: String, message: String, signature: String): Boolean {
        if (secret.isEmpty() && signature.isEmpty()) return true
        return sign(secret, message) == signature
    }

    fun buildSigningMessage(method: String, path: String, timestamp: String, body: String): String {
        return "$method:$path:$timestamp:$body"
    }
}
